### PR TITLE
p2p: Remove GetAdjustedTime() from AddrMan

### PR DESCRIPTION
**Pull request by w0xlt**
This PR modifies AddrMan to use local time instead of `GetAdjustedTime()`.

Motivation: Check discussion in https://github.com/bitcoin/bitcoin/pull/23695, https://github.com/bitcoin/bitcoin/pull/23631 and the issues related to `GetAdjustedTime()` reported in https://github.com/bitcoin/bitcoin/issues/4521

### CODE REVIEW

**Comment by laanwj** [comment]
Concept ACK

**Comment by w0xlt** [comment]
@laanwj changed the last commit (f282ffb) to a scripted-diff.

**Comment by MarcoFalke** [comment]
I haven't thought a lot about this, but I have two questions:

Is this going to cause any additional fingerprinting vectors? I guess not after commit 14ba286556faad794f288ef38493c540382897cb?

Also, could this break P2P address relay? I guess only for the local node?

**Comment by laanwj** [comment]
> Also, could this break P2P address relay? I guess only for the local node?

Yes, but so could a misadjusted time due to connecting to the wrong nodes. At the least it reduces variance between starts of `bitcoind`.

**Comment by pawan17kumar** [comment]
Why you sending mails, just stop it.

**Comment by mzumsande** [comment]
> Also, could this break P2P address relay?

This PR doesn't affect address (gossip) relay, because that is independent from AddrMan:
The decision whether to forward an address will still be on an adjusted time basis after this, see https://github.com/bitcoin/bitcoin/blob/cb27d60f966ad20c465e5db05ba1a2253bd4dab3/src/net_processing.cpp#L2887-L2888
and 
https://github.com/bitcoin/bitcoin/blob/cb27d60f966ad20c465e5db05ba1a2253bd4dab3/src/net_processing.cpp#L2933

Although it would probably make sense to change `net_processing` too as well to be consistent.

**Comment by naumenkogs** [comment]
Could you summarize the reasons for this change and put it in the OP post?
It's difficult to make sure everyone's on the same page otherwise.

**Comment by w0xlt** [comment]
@naumenkogs The main reason is to remove the use of GetAdjustedTime.

As @mzumsande said, this change does not affect the network layer (or other layers), only the local AddrMan so it seems safer and easier to test and review.

I have a version of this PR that also removes `GetAdjustedTime` from `src/node/miner.cpp`, `src/validation.cpp`, `src/net.cpp`, `src/rpc/net.cpp` and `src/net_processing.cpp`.  While this passes the tests (without changing them), I'm not sure what behavioral changes this might cause.

If reviewers think it's better, I can update the PR with this version.

File: src/timedata.h [hunk 1]
```diff
@@ -73,6 +73,7 @@ class CMedianFilter
 /** Functions to keep track of adjusted P2P time */
 int64_t GetTimeOffset();
 int64_t GetAdjustedTime();
+int64_t GetLocalTime();
```
Comment (MarcoFalke) [inline src/timedata.h:76]: what is this? I'd say please use `GetTime` or (if possible) `GetTime<chrono>`

**Comment by MarcoFalke** [comment]
> If reviewers think it's better, I can update the PR with this version.

If this really doesn't change p2p behavior and is only a refactor (?), then I'd say to leave this pull as-is to simplify review. If this is not a refactor, it would be good to list all behavior changes.

**Comment by naumenkogs** [comment]
@mzumsande I'm not sure this change doesn't affect p2p

1) For example, we send addrs to our peers via ADDRv2 message, and those contain timestamps, which mean something for the peer. (Bitcoin Core might be using those very mildly, but anyway, it does look at them).
2) Also, making it further from mainstream (not using adjusted time) might add more fingerprinting capabilities by looking at those timestamps

@w0xlt 

And what's the problem with `GetAdjustedTime` in addrman exactly?

**Comment by MarcoFalke** [comment]
> And what's the problem with GetAdjustedTime in addrman exactly?

The problem is that adjusted time is:
* controlled by "the network"
* it is not guaranteed to be monotonic
* it doesn't even guarantee to be the correct time (or close to it), even if all your peers are well-behaved

Thus, it is an huge mental burden to think about any change to addrman that involves those timestamps. Also, it makes it hard to understand the existing logic, even if nothing is changed.

**Comment by mzumsande** [comment]
> @mzumsande I'm not sure this change doesn't affect p2p

I wasn't trying to say that, just that it doesn't really affect gossip relay, because self-announcements have already been changed to always use local time in #23695, and the mechanism of whether to forward these is located in `net_processing` , not in AddrMan.

One p2p aspect that the adjusted time in AddrMan can influence in theory are GETADDR responses:

Addrs for which `IsTerrible()` is true are not included in these, and the Adjusted Time influences `IsTerrible()` both directly as a parameter, and indirectly, by influencing `nLastTry`, `nLastSuccess` and `nAttempts`, `nTime` of an address, which are updated by functions using the adjusted time (such as `Connected()`, `Good()` etc.).

I'm not sure that this is a huge influence though, since `IsTerrible()` mostly operates on larger timescales (days instead of minutes), GETADDR responses are calculated once a day and cached - whereas the adjusted time would differ from the local time only slightly when they deviate.

I'd have to think about it a bit more, but I tend toward removing the Adjusted Time from all addr related code - not just AddrMan but also in `net_processing`.

**Comment by naumenkogs** [comment]
@mzumsande 

>the adjusted time would differ from the local time only slightly when they deviate.

What is the biggest possible deviation? From what I see in the code, any, but we show warnings if it deviates more than 5 minutes? I thought the deviation from block header timestamps is limited, but I can't find that.

>I'd have to think about it a bit more, but I tend toward removing the Adjusted Time from all addr related code - not just AddrMan but also in net_processing.

I probably support the overall sentiment, I just think we should understand all implications and document it here.
E.g., what do you think about the privacy leak I mentioned? Do you agree it exists, and consider it too mild to pose a threat?

**Comment by MarcoFalke** [comment]
> What is the biggest possible deviation?

Likely `-maxtimeadjustment`?

File: src/timedata.h (hunk 1 above)
Comment (w0xlt) [reply src/timedata.h]: The idea of `GetLocalTime` was to encapsulate `count_seconds(GetTime<std::chrono::seconds>()` to avoid repetition.

I removed GetLocalTime` function, but `GetTime<chrono>` can't be used without `count_seconds` since `GetAdjustedTime` returns `int_64`.

**Comment by w0xlt** [comment]
> If this really doesn't change p2p behavior and is only a refactor (?),

This is not just a refactor because it changes the AddrMan behavior.
But since @mzumsande suggests changing `net_processing` as well, I added this change.

> If this is not a refactor, it would be good to list all behavior changes.

commit 3a141ff

1. `AddrMan::Good()` function will use (by default) the current time (instead of the adjusted time) to update the time when the node last connected to a peer (`AddrInfo::nLastSuccess`,  `AddrInfo::nLastTry` and `AddrManImpl::nLastGood`) (`addrman.h:92`).

2. `AddrMan::Attempt()` function will use (by default) the current time (instead of the adjusted time) to update the `AddrInfo::nLastTry` (`addrman.h:95`).

3. `AddrManImpl::AddSingle()` will use the current time (instead of the adjusted time) to update `CAddress::nTime` periodically (`addrman.cpp:559`).

4. `AddrManImpl::GetAddr_()` will use the current time (instead of the adjusted time) as a filter for `AddrInfo::IsTerrible()` (`addrman.cpp:785`)

5. AddrManImpl::ResolveCollisions_() will use the current time (instead of the adjusted time) to compare `AddrInfo::nLastSuccess` or `AddrInfo::nLastTry`. Note that these fields are also updated with the current time in this PR (items 1 and 2). (`addrman.cpp:871-892`)

6. `AddrInfo::IsTerrible()` will use the current time (instead of the adjusted time) to determine whether the entry are bad enough to be deleted (`addrman_impl:94`).

7. `AddrInfo::GetChance()` will use the current time (instead of the adjusted time) to calculate `nSinceLastTry` to deprioritize very recent attempts (`addrman_impl:97`).

commit cb660ce

1. `PeerManagerImpl::CanDirectFetch()` will use the current time (instead of the adjusted time). This function is used to set the flag for initiating extra block-relay-only peer connections, to allow the creation of CMPCTBLOCK message and others. (`net_processing:960`).

2. `PeerManagerImpl::ProcessMessage()` will use the current time (instead of the adjusted time) to store the new addresses when receiving ADDR or ADDRV2 messages. (`net_processing:2889`)

3. `PeerManagerImpl::MaybeSendAddr` will use the current time (instead of the adjusted time)  to calculate if the block time is close to today (`net_processing:4599`), the `CNodeState::m_headers_sync_timeout` (`net_processing:4606`) and to detect if a initial-headers-sync peer is unresponsive (`net_processing:4640`)

**Comment by josibake** [comment]
Concept ACK

I have found understanding `GetAdjustedTime()` to be a large stumbling block as I've been reading through `AddrMan`.

I'll spend some time reading the linked threads and discussion on this PR before digging in a bit more

File: src/addrman.cpp [hunk 2]
```diff
@@ -868,28 +868,28 @@ void AddrManImpl::ResolveCollisions_()
...
                     // just evict the old entry -- we must not be able to
                     // connect to it for some reason.
                     LogPrint(BCLog::ADDRMAN, "Unable to test; replacing %s with %s in tried table anyway\n", info_old.ToString(), info_new.ToString());
-                    Good_(info_new, false, GetAdjustedTime());
+                    Good_(info_new, false, count_seconds(GetTime<std::chrono::seconds>()));
                     erase_collision = true;
                 }
             } else { // Collision is not actually a collision anymore
-                Good_(info_new, false, GetAdjustedTime());
+                Good_(info_new, false, count_seconds(GetTime<std::chrono::seconds>()));
```
Comment (MarcoFalke) [inline src/addrman.cpp:892]: `count_seconds` should only be used to *serialize* the time value in seconds. For example for RPC, the debug log, or the GUI, ...

It shouldn't be used for normal code. You can simply use the equivalent `GetTime` here. If you want to use std::chrono, you should update all related types, but this should be done in a separate commit. Ideally a separate pull request.

Comment (w0xlt) [reply src/addrman.cpp]: Changed to  `GetTime`.

**Comment by w0xlt** [comment]
Rebased.

File: src/net_processing.cpp [hunk 3]
```diff
@@ -2891,7 +2891,7 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
 
         // Store the new addresses
         std::vector<CAddress> vAddrOk;
-        int64_t nNow = GetAdjustedTime();
+        int64_t nNow = GetTime();
```
Comment (naumenkogs) [inline src/net_processing.cpp:2894]: One side effect I’m thinking of is `if (addr.nTime <= 100000000 || addr.nTime > nNow + 10 * 60)`. If this is triggered (because our node lives in the past),  we would mark some of the recent address (including all one-hop self-announcements) to be 5 days old, and deprioritize them?

**Review by maflcko — COMMENTED** [review]
Looks like this is changing both too much (a few in net_processing) and too little (missing net.cpp).

**Comment by MarcoFalke** [comment]
Thanks, but due to inactivity I am closing this for now. I've picked it up in #24662.
